cloudinit/config/cc_ubuntu_advantage.py

```diff
@@ -279,16 +372,19 @@ def handle(
         raise RuntimeError(msg)
 
     config = ua_section.get("config")
-
     if config is not None:
         supplemental_schema_validation(config)
 
     maybe_install_ua_tools(cloud)
-    configure_ua(
-        token=ua_section.get("token"),
-        enable=ua_section.get("enable"),
-        config=config,
+
+    # ua-auto-attach.service had noop-ed as ua_section is not empty
+    disable_auto_attach = bool(
+        ua_section.get("features", {}).get("disable_auto_attach", False)
     )
+    if not disable_auto_attach and _should_auto_attach():
+        _auto_attach(ua_section)
+    elif not ua_section.keys() <= {"features"}:
```

Comment:
Something seems amiss here. Please clarify the logic here for me.

The paths written seem to be:
1. call _auto_attach if  not `ubuntu_advantage: {features: {disable_auto_attach: true}}` AND on PRO image
2. call _attach only if (disable_auto_attach OR not PRO) AND ua_section has top-level keys besides "features"

So, if non-PRO images provide "ubuntu_advantage: features" beyond the `disable_auto_attach` allowed in our schema, they'll get a schema warning `Invalid cloud-config provided: Please run 'sudo cloud-init schema --system'` because of denied features properties.

Did we decide that we will be waiting on UA python api to support for passing any feature flag from cloud-init? 

Maybe cloud-init schema allowing for direct passthrough of any uaclient.conf `features` can wait the presence of UA's python API with support for setting supplemental cloud-init requested features before trying to permit that in `#cloud-config` schema.

I think for customers that want to set random UA client features today via `#cloud-config` user-data, can alternatively provide a `write_files` section like the following and cloud-init schema may not necessarily have to define and support to passthrough any uaclient.conf features flags via cloud-init's schema:
```
# Set UA feature flag for UA client before attach or auto-attach is performed by cloud-init
#cloud-config  
write_files:
- path: /etc/ubuntu-advantage/uaclient.conf
  append: true
  content: |
    features:
      some_feature_name: true
```


Maybe it's worth documenting this type of write_files example in our cc_ubuntu_advantage docs to give folks an example of trying to augment feature flags in uaclient.conf

@CalvoM do we have any backlog items in ua-client API that would eventually surface the ability to set `features` overrides for uaclient via cloud-init or are we always expecting that any behavior overrides are provided directly to the UA api operation like auto_attach_full vs short. Right now only `enable` and `enable_beta` options.

@aciba90 what do you think of in the short-term just documenting the above example in cc_ubuntu_advantage.meta.examples just so folks looking to further customize UA behavior have a starting point to work with.

Reply:
> Something seems amiss here. Please clarify the logic here for me.
>
> The paths written seem to be:
>
>   1. call _auto_attach if not ubuntu_advantage: {features: {disable_auto_attach: true}} AND on PRO image
>   2. call _attach only if (disable_auto_attach OR not PRO) AND ua_section has top-level keys besides "features"

The first branch is for UA Pro configurations, and the second branch supports the following cases:

1. Attach configurations. i.e. a configuration with `{"token": "<token>", "enable": ["esm"]}`
2. If the configuration only contains the sub-property `features` or it is empty, then do not perform `attach`. It could sound redundant, but a user could want just to disable auto-attach and not perform an attach with the following configuration: `{"features": {"disable_auto_attach": True}}`.

I have written a comment explaining it and renamed plus extended the unit-test [test_cc_ubuntu_advantage.py::TestHandle::test_handle_auto_attach_vs_attach](https://github.com/canonical/cloud-init/pull/1583/files#diff-c588a33656e0018aed4e9aa9b4e90815b6dcd762595ecaf6da132cf9ed681382R681) which cover these cases.

> So, if non-PRO images provide "ubuntu_advantage: features" beyond the disable_auto_attach allowed in our schema, they'll get a schema warning Invalid cloud-config provided: Please run 'sudo cloud-init schema --system' because of denied features properties.

I have defined the `features` object as `additionalProperties: true` to allow that case and we will log a debug message from [cc_ubuntu_advantage::validate_schema_features](https://github.com/canonical/cloud-init/pull/1583/files#diff-e12ed348cca4b27232907df295029b93b250cec08d438f0ab37e498a6df3a263R141) if extra features are given.

> Did we decide that we will be waiting on UA python api to support for passing any feature flag from cloud-init?

I think we concluded that UA features' configuration from cloud-init was out of the scope of this roadmap item.

> @aciba90 what do you think of in the short-term just documenting the above example in cc_ubuntu_advantage.meta.examples just so folks looking to further customize UA behavior have a starting point to work with.

An argument against this was exposed during our [discussions](https://docs.google.com/document/d/1tc3ZJC4wD6iyMWCFXC_rxWzbjrl0mN4J_YbF2iFoiRY/edit?pli=1&disco=AAAAbOEBxfY). Changes to this file could provoke conflicts during the upgrade of `ubuntu-advantage-tools`.

Another argument against it is that those features are not documented by UA. That makes me think whether UA is going to maintain backwards compatibility of them or not.

I think it would not hurt to add the example, but making clear that those features are an advanced topic and that modifying that file could cause friction during system upgrades.

Reply:
> I think we concluded that UA features' configuration from cloud-init was out of the scope of this roadmap item.

Great thanks for the clarifications

>  I have defined the features object as additionalProperties: true to allow that case and we will log a debug message from [cc_ubuntu_advantage::validate_schema_features](https://github.com/canonical/cloud-init/pull/1583/files#diff-e12ed348cca4b27232907df295029b93b250cec08d438f0ab37e498a6df3a263R141) if extra features are given.

@aciba90 please set back the  [`additionalProperties: false`](https://github.com/canonical/cloud-init/pull/1583/files#diff-5092062d78cf6dbc9133596b51c11590df10b46d68c7257257486b2911f05b6dR2221) for features and drop the "[Ignoring unknown  'ubuntu_advantage.features loggiing](https://github.com/canonical/cloud-init/pull/1583/files#diff-e12ed348cca4b27232907df295029b93b250cec08d438f0ab37e498a6df3a263R159). The schema error messaging in this case is clear enough and it just doesn't make sense to allow providing unsupported `features` in cloud-config only to log that we are ignoring them.

```bash
$ PYTHONPATH=. python3 -m cloudinit.cmd.main schema --annotate -c ua.yaml

#cloud-config
ubuntu_advantage:
 token: Redacted
 enable_beta: ['str']
 features:		# E1
   a: nope

# Errors: -------------
# E1: Additional properties are not allowed ('a' was unexpected)
```


> An argument against this was exposed during our [discussions](https://docs.google.com/document/d/1tc3ZJC4wD6iyMWCFXC_rxWzbjrl0mN4J_YbF2iFoiRY/edit?pli=1&disco=AAAAbOEBxfY). Changes to this file could provoke conflicts during the upgrade of ubuntu-advantage-tools.

Good point. Ok, best not to publicly document something which could be hard to maintain from cloud-init and or UA in the future. 

> I think it would not hurt to add the example, but making clear that those features are an advanced topic and that modifying that file could cause friction during system upgrades.

Your points carry good weight here. Let's avoid the extra documentation on setting other features with `write_files`. Any customer that needs these features will be told specifically about them and we don't want to get into a support/upgrade debacle here with something that isn't a primary mechanism to configure/run UA.

Reply:
Done, thanks.